packages/@aws-cdk/pipelines/lib/private/application-security-check.ts

```diff
@@ -0,0 +1,112 @@
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id);
+
+    this.preApproveLambda = new lambda.Function(this, 'CDKPipelinesAutoApprove', {
+      handler: 'index.handler',
+      runtime: lambda.Runtime.PYTHON_3_8,
+      code: lambda.Code.fromAsset(path.resolve(__dirname, '../lambda')),
+      timeout: Duration.seconds(30),
+    });
+
+    this.preApproveLambda.addToRolePolicy(new iam.PolicyStatement({
+      actions: ['codepipeline:GetPipelineState', 'codepipeline:PutApprovalResult'],
+      conditions: {
+        StringEquals: {
+          'aws:ResourceTag/SECURITY_CHECK': 'ALLOW_APPROVE',
+        },
+      },
+      resources: ['*'],
+    }));
+
+    const assemblyPath = 'assembly-$STACK_NAME-$STAGE_NAME/';
+    const invokeLambda =
+      'aws lambda invoke' +
+      ` --function-name ${this.preApproveLambda.functionName}` +
+      ' --invocation-type Event' +
+      ' --payload "$payload"' +
+      ' lambda.out';
+
+    this.cdkDiffProject = new codebuild.Project(this, 'CDKSecurityCheck', {
+      buildSpec: codebuild.BuildSpec.fromObject({
+        version: 0.2,
+        phases: {
+          build: {
+            commands: [
+              'npm install -g aws-cdk',
+              // $CODEBUILD_INITIATOR will always be Code Pipeline and in the form of:
+              // "codepipeline/example-pipeline-name-Xxx"
+              'payload="$(node -pe \'JSON.stringify({ "PipelineName": process.env.CODEBUILD_INITIATOR.split("/")[1], "StageName": process.env.STAGE_NAME, "ActionName": process.env.ACTION_NAME })\' )"',
+              // ARN: "arn:aws:codebuild:$region:$account_id:build/$project_name:$project_execution_id$"
+              'ARN=$CODEBUILD_BUILD_ARN',
+              'REGION="$(node -pe \'`${process.env.ARN}`.split(":")[3]\')"',
+              'ACCOUNT_ID="$(node -pe \'`${process.env.ARN}`.split(":")[4]\')"',
+              'PROJECT_NAME="$(node -pe \'`${process.env.ARN}`.split(":")[5].split("/")[1]\')"',
+              'PROJECT_ID="$(node -pe \'`${process.env.ARN}`.split(":")[6]\')"',
+              // Manual Approval adds 'http/https' to the resolved link
+              'export LINK="$REGION.console.aws.amazon.com/codesuite/codebuild/$ACCOUNT_ID/projects/$PROJECT_NAME/build/$PROJECT_NAME:$PROJECT_ID/?region=$REGION"',
+              // Run invoke only if cdk diff passes (returns exit code 0)
+              // 0 -> true, 1 -> false
+              `if cdk diff -a "${assemblyPath}" --security-only --fail; then ${invokeLambda}; export MESSAGE='No changes detected.'; ` +
+              'else export MESSAGE="Changes detected! Requires Manual Approval"; ' +
+              'fi',
+            ],
+          },
+        },
+        env: {
+          'exported-variables': [
+            'LINK',
+            'MESSAGE',
+          ],
+        },
+      }),
+    });
+
+    this.cdkDiffProject.addToRolePolicy(new iam.PolicyStatement({
+      actions: ['cloudformation:DescribeStacks', 'cloudformation:GetTemplate'],
```

Comment:
Instead of these permissions, the diff project actually needs `sts:AssumeRole` on the deploy role. I wonder if writing this will do it (except replace the ARN with a `*`, don't to `Stack.of(this).account` because if you do, it won't work in other accounts):

https://github.com/aws/aws-cdk/blob/94f81c441f9e2cb8dc70eb2e772d2cd75e468b67/packages/%40aws-cdk/pipelines/lib/actions/update-pipeline-action.ts#L111-L119

This is the only correct way to make it work cross-account.